### PR TITLE
Refactor QP with MatrixOfConstraints and bridges

### PR DESCRIPTION
**Pull request by blegat**
Requires

- [x] https://github.com/jump-dev/MathOptInterface.jl/issues/1711
- [x] Create proper PRs with tests of the needed changes in https://github.com/jump-dev/MathOptInterface.jl/tree/bl/diffopt
- [x] https://github.com/jump-dev/MathOptInterface.jl/pull/1716
- [x] https://github.com/jump-dev/MathOptInterface.jl/pull/1718
- [x] Replace the checkout of `MOI#bl/diffopt` in the Github actions to `MOI#master`.
- [x] Release MOI v0.10.8
- [x] Remove the checkout of `MOI#master`.
- [x] Add MOI v0.10.8 as a requirement in the Project.toml
- [x] Find what to do with affine variable bridges (`bridged_function` and `unbridged_function` don't treat the function as a derivative hence they will add constants :/)

Closes https://github.com/jump-dev/DiffOpt.jl/issues/153
Closes https://github.com/jump-dev/DiffOpt.jl/issues/167

### CODE REVIEW

**Comment by matbesancon** [comment]
Current failures are mostly due to:
```
ArgumentError: Ipopt.Optimizer does not support getting the attribute MathOptInterface.ListOfModelAttributesSet().
```

**Comment by blegat** [comment]
> Current failures are mostly due to:
> 
> ```
> ArgumentError: Ipopt.Optimizer does not support getting the attribute MathOptInterface.ListOfModelAttributesSet().
> ```

Fixed by 93d286c

File: src/moi_wrapper.jl [hunk 1]
```diff
@@ -21,7 +21,17 @@ julia> _backward_quad(model)  # for convex conic models
...
+    # When we do `MOI.copy_to(diff, optimizer)` we need to efficiently `MOI.get`
+    # the model information from `optimizer`. However, 1) `optimizer` may not
+    # implement some getters or it may be inefficient and 2) the getters may be
+    # unimplemented or inefficient through some bridges.
+    # For this reason we add a cache layer, the same cache JuMP adds.
+    caching_opt = MOIU.CachingOptimizer(
+        MOIU.UniversalFallback(MOIU.Model{Float64}()),
+        optimizer,
+    )
+    return Optimizer(caching_opt)
```
[the remark this answers is not in the document]
Comment (blegat) [reply src/moi_wrapper.jl:34]: We should recommend `DiffOpt.diff_optimizer` in both cases. `Optimizer` is for advanced user that want to control the caches and bridges, similar to the direct mode in JuMP.

**Comment by matbesancon** [comment]
Is this blocked by a merge on MOI? I saw the special branch in the docs

**Comment by blegat** [comment]
> Is this blocked by a merge on MOI? I saw the special branch in the docs

Yes, the MOI `bl/diffopt` contains random changes needed. It should be separated into documented and tested changes, merged to MOI and then MOI should get a new release.
https://github.com/jump-dev/MathOptInterface.jl/pull/1716 is the first one of the list.

**Comment by matbesancon** [comment]
@blegat we can now change the dependency to MOI 0.10.8 and remove the branch specification in CI :)

**Comment by matbesancon** [comment]
>    Evaluated: [0.9998383009322291] ≈ [1.0] (atol=0.0001, rtol=0.0001)

@blegat I would just crank up the tolerance here no?

**Pull request merged**